tensorflow_io/core/kernels/http_kernels.cc

```diff
@@ -13,16 +13,527 @@ See the License for the specific language governing permissions and
+      TF_SetStatus(status, TF_INTERNAL, error_message.c_str());
+      return;
+    }
+
+    if ((s = curl_easy_setopt(curl_, CURLOPT_XFERINFOFUNCTION,
+                              &CurlHttpRequest::ProgressCallback)) !=
+        CURLE_OK) {
+      std::string error_message =
+          absl::StrCat("Unable to set CURLOPT_XFERINFOFUNCTION: ", s);
+      TF_SetStatus(status, TF_INTERNAL, error_message.c_str());
+      return;
+    }
+
+    response_buffer_.reserve(CURL_MAX_WRITE_SIZE);
+    if ((s = curl_easy_setopt(curl_, CURLOPT_WRITEDATA,
+                              reinterpret_cast<void*>(this))) != CURLE_OK) {
+      std::string error_message =
+          absl::StrCat("Unable to set CURLOPT_WRITEDATA: ", s);
+      TF_SetStatus(status, TF_INTERNAL, error_message.c_str());
+      return;
+    }
+    if ((s = curl_easy_setopt(curl_, CURLOPT_WRITEFUNCTION,
+                              &CurlHttpRequest::WriteCallback)) != CURLE_OK) {
+      std::string error_message =
+          absl::StrCat("Unable to set CURLOPT_WRITEFUNCTION ", s);
+      TF_SetStatus(status, TF_INTERNAL, error_message.c_str());
+      return;
+    }
+
+    TF_SetStatus(status, TF_OK, "");
+  }
+
+  void SetUri(const std::string& uri, TF_Status* status) {
+    CURLcode s = CURLE_OK;
+    if ((s = curl_easy_setopt(curl_, CURLOPT_URL, uri.c_str())) != CURLE_OK) {
+      std::string error_message =
+          absl::StrCat("Unable to set CURLOPT_URL (", uri, "): ", s);
+      TF_SetStatus(status, TF_INTERNAL, error_message.c_str());
+      return;
+    }
+
+    uri_ = uri;
+
+    TF_SetStatus(status, TF_OK, "");
+  }
+
+  void SetRange(uint64_t start, uint64_t end, TF_Status* status) {
+    CURLcode s = CURLE_OK;
+    if ((s = curl_easy_setopt(curl_, CURLOPT_RANGE,
+                              absl::StrCat(start, "-", end).c_str())) !=
+        CURLE_OK) {
+      std::string error_message = absl::StrCat("Unable to set CURLOPT_RANGE (",
+                                               start, "-", end, "): ", s);
+      TF_SetStatus(status, TF_INTERNAL, error_message.c_str());
+      return;
+    }
+
+    TF_SetStatus(status, TF_OK, "");
+  }
+
+  void SetResultBufferDirect(char* buffer, size_t size, TF_Status* status) {
+    direct_response_ = DirectResponseState{buffer, size, 0, 0};
+    CURLcode s = CURLE_OK;
+    if ((s = curl_easy_setopt(curl_, CURLOPT_WRITEDATA,
+                              reinterpret_cast<void*>(this))) != CURLE_OK) {
+      std::string error_message =
+          absl::StrCat("Unable to set CURLOPT_WRITEDATA: ", s);
+      TF_SetStatus(status, TF_INTERNAL, error_message.c_str());
+      return;
+    }
+    if ((s = curl_easy_setopt(curl_, CURLOPT_WRITEFUNCTION,
+                              &CurlHttpRequest::WriteCallbackDirect)) !=
+        CURLE_OK) {
+      std::string error_message =
+          absl::StrCat("Unable to set CURLOPT_WRITEDATA: ", s);
+      TF_SetStatus(status, TF_INTERNAL, error_message.c_str());
+      return;
+    }
```

Comment:
@yongtang these two conditions are covered in `Initialize()`. Any specific reason for adding here again?

Reply:
@kvignesh1420 There are two cases. One is to set the result buffer with caller provided buffer (`SetResultBufferDirect`). Another is the non-caller provided buffer. To make it clear I have split out the part with new function `SetResultBuffer`.